[LV] Fix FindLastIV reduction for epilogue vectorization.

Mel-Chen opened the pull request:
Following 0e528ac404e13ed2d952a2d83aaf8383293c851e, this patch adjusts the resume value of VPReductionPHIRecipe for FindLastIV reductions. Replacing the resume value with:
```
ResumeValue = ResumeValue == StartValue ? SentinelValue : ResumeValue;
```
This addressed the correctness issue when the start value might not be less than the minimum value of a monotonically increasing induction variable.

Thanks Florian Hahn for the help.

Mel-Chen committed:
[LV] Fix FindLastIV reduction for epilogue vectorization.

fhahn commented on llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp:
```diff
@@ -3405,15 +3406,13 @@ void VPReductionPHIRecipe::execute(VPTransformState &State) {
     // a constant, and smaller than the start value of the induction variable,
     // the start value can be directly used to initialize the reduction phi.
-    Iden = StartV;
+    StartV = Iden = RdxDesc.getSentinelValue();
```
Is there a reason for ignoring the start value of the recipe here? Would be good to understand why we need to handle this differently than other reduction resume values (and all other phi recipes)

Mel-Chen replied:
This is for correctness. For example: 
```
int rdx = 3;
for (int i = 0; i < n; ++i) 
  rdx = a[i] > 3 ? i : rdx;
return rdx;
```
Assume {0, +, 1} generates a sequence <0, 1, 2, 3, ..., n-1>. This sequence might include the start value (3 in the example), so we need a number (the sentinel) that does not appear in this sequence and is smaller than all its members to replace the start value during the reduction operation. After the reduction, the result can be corrected using `rdx_res = max_rdx != sentinel ? max_rdx : start_value`.

Back to epilogue vectorization, the same approach must be applied to implement FindLastIV, especially since the main vector loop may be skipped. In such cases, the resume value entering the epilogue vector loop will directly be the start value. The epilogue vector loop must handle this in the same way as the main vector loop.

Interestingly, in most real-world code, it probably looks like this: 
```
int rdx = -1;
for (int i = 0; i < n; ++i) 
  rdx = a[i] > 3 ? i : rdx;
return rdx;
```
In this case, directly initializing the phi with the start value (-1 in the second example) is entirely correct. As long as the start value does not appear in the {0, +, 1} sequence and is smaller than all the sequence members, a sentinel is unnecessary, and the start value can be used to initialize the phi directly. This is a improvement  direction for FindLastIV idiom.

fhahn replied:
Would it be possible to do something similar to how we handle the start values for AnyOf in the epilogue vector loop (https://github.com/llvm/llvm-project/blob/main/llvm/lib/Transforms/Vectorize/LoopVectorize.cpp#L10272), i.e. could we create the new start value for the epilogue loop by checking if the value out of the main vector loop is still the sentinel value or something like that

Mel-Chen replied:
AnyOf can do this because there is only 1 element in the sequence.
FindLastIV possible needs to be adjusted like this:
Change bc.merge.rdx from
```
; CHECK-NEXT: [[BC_MERGE_RDX:%.*]] = phi i64 [ [[RDX_SELECT]], %[[VEC_EPILOG_ITER_CHECK]] ], [ 3, %[[VECTOR_MAIN_LOOP_ITER_CHECK]] ]
```
to 
```
; CHECK-NEXT: [[BC_MERGE_RDX:%.*]] = phi i64 [ [[TMP6]], %[[VEC_EPILOG_ITER_CHECK]] ], [ -9223372036854775808, %[[VECTOR_MAIN_LOOP_ITER_CHECK]] ]
;; TMP6 is defined by `[[TMP6:%.*]] = call i64 @llvm.vector.reduce.smax.v4i64(<4 x i64> [[TMP4]])`
```
The method is to pattern match `rdx_res = max_rdx != sentinel ? max_rdx : start_value`, and directly use the result of reduce.smax (i.e. `max_rdx`) to resume the reduction result instead of the adjusted result (i.e. `rdx_res`).

I think this is a little more complicated, what do you think?

Mel-Chen replied:
I've given it more thought and realized there might be a more concise approach!
Perhaps we could modify ResumeV as:
```
ResumeV = ResumeV == StartValue ? Sentinel : ResumeV
```
I'll try to make this change and test it tomorrow to see if there are any blind spots we have not notice.

fhahn replied:
Yes this is along the lines of what I was thinking. It would be great if we could convey all necessary info via the start value of the recipe (in line with all other reduction types) instead of ignoring it

fhahn replied:
Great, that was what I had in mind :)

fhahn commented:
It would be good to have more runtime test coverage for the FindLastIV logic, could you add some end-to-end test to `llvm-test-suite`, possibly similar to https://github.com/llvm/llvm-test-suite/blob/main/SingleSource/UnitTests/Vectorizer/index-select.cpp

Mel-Chen committed:
Revert "[LV] Fix FindLastIV reduction for epilogue vectorization."

This reverts commit be9d783ba941a80896f861ff424d81b886680615.

Mel-Chen commented:
> It would be good to have more runtime test coverage for the FindLastIV logic, could you add some end-to-end test to `llvm-test-suite`, possibly similar to https://github.com/llvm/llvm-test-suite/blob/main/SingleSource/UnitTests/Vectorizer/index-select.cpp

Sure, I should. https://github.com/llvm/llvm-test-suite/pull/193

Mel-Chen committed:
Revert "Revert "[LV] Fix FindLastIV reduction for epilogue vectorization.""

This reverts commit 895a134e768ae2467016d9e3e3641b0e92b4a3fa.

Mel-Chen committed:
Revert "[LV] Fix FindLastIV reduction for epilogue vectorization."

This reverts commit 7e233ba712e8842ab65dbddeedb5fa82cf1dfc03.

Mel-Chen committed:
Adjust resume value of FindLastIV

Use
ResumeV = ResumeV == StartValue ? Sentinel : ResumeV
to adjust the resume value.

Mel-Chen commented:
The llvm test-suite result of current implementation:
```
Checking findlast_true_update
Checking findlast_false_update
Checking findlast_start_TC
Miscompare
```
The fail can be fixed after this patch:
```
Checking findlast_true_update
Checking findlast_false_update
Checking findlast_start_TC
Checking findlast_inc_2
Checking findlast_start_decreasing_induction
Checking findlast_iv_start_3
Checking findlast_start_3_iv_start_3
Checking findlast_start_2_iv_start_3
Checking findlast_start_4_iv_start_3
```
The test config is `-mcpu=sifive-x280 -O3 -mllvm -epilogue-vectorization-force-VF=2`.

fhahn commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -7687,6 +7687,19 @@ static void fixReductionScalarResumeWhenVectorizingEpilog(
+        match(Cmp,
+              m_SpecificICmp(ICmpInst::ICMP_EQ, m_Specific(OrigResumeV),
+                             m_Specific(RdxDesc.getRecurrenceStartValue())));
+    assert(IsExpectedPattern && "Unexpected reduction resume pattern");
```
```suggestion
    assert(IsExpectedPattern && "Unexpected reduction resume pattern");
    (void) IsExpectedPattern;
```

to avoid warning on release builds

fhahn reviewed (approved):
LGTM, thanks!

Mel-Chen committed:
Update llvm/lib/Transforms/Vectorize/LoopVectorize.cpp

Avoid warning on release build

Co-authored-by: Florian Hahn <flo@fhahn.com>

Mel-Chen committed:
Format
